rootmap/sqlrootmap.go

```diff
@@ -65,23 +201,48 @@ func (rm *sqlRootmap) Put(
 ) error {
 	rm.mtx.Lock()
 	defer rm.mtx.Unlock()
-	_, err := rm.db.ExecContext(ctx, `
-	insert into rootmap (
-		database,
-		producer_id,
-		topic,
-		version,
-		storage_prefix,
-		node_id
-	) values ($1, $2, $3, $4, $5, $6)`,
-		database, producerID, topic, version, prefix, hex.EncodeToString(nodeID[:]),
-	)
-	if err != nil {
-		var err sqlite3.Error
-		if errors.As(err, &sqlite3.Error{Code: sqlite3.ErrConstraint}) {
-			return ErrRootAlreadyExists
+	err := rm.withTransaction(func(tx *sql.Tx) error {
+		var tableID int
+		err := rm.db.QueryRowContext(ctx, `
+		select id from tables where producer = ? and topic = ? and database = ? 
+		`, producerID, topic, database).Scan(&tableID)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("failed to read from tables: %w", err)
 		}
-		return fmt.Errorf("failed to store to rootmap: %w", err)
+		if errors.Is(err, sql.ErrNoRows) {
+			err := rm.db.QueryRowContext(ctx, `
+			insert into tables (database, producer, topic) values (?, ?, ?)
+			returning id
+			`, database, producerID, topic).Scan(&tableID)
+			if err != nil {
+				return fmt.Errorf("failed to insert into tables: %w", err)
+			}
+			rm.tables[table{database, producerID, topic}] = true
+		}
+
+		timestamp := time.Now().UTC().Format(time.RFC3339Nano)
+
+		_, err = rm.db.ExecContext(ctx, `
```

Comment:
I think you mean for this to be executed in the transaction, so should be

```
tx.ExecContext(ctx, ...)
```

likewise on 213 and anywhere inside these withTransaction blocks.